LIN_ADVANCE updates

Sebastianv650 opened the pull request:
Some updates for LIN_ADVANCE:

.) long to int for Stepper::e_steps: As far as I know @esenapaj changed it to long before, but at last for me it's to slow this way. My extruder stalls at 10kHz / 20kHz step limits with long. Maybe int is OK now for everyone due to the implemented STEP_PULSE_CYCLES?
.) Take the delta_adv_steps calculation out of the step_loops loop. Wasted calculation performance if done inside.
.) >> 2 replaced by 3 for the eISR_Rate calculation: I'm not shure where the 2 comes from. Maybe a typo on my side in the version I used for the first pull requests.. Reason: Timer 1 runs at 16/8=2MHz, Timer 0 at 16/64=0.25MHz. ==> 2/0.25=8.

Everybody feels fine with the changes?

Commit:
Merge pull request #2 from MarlinFirmware/RCBugFix

Rc bug fix

Commit:
Merge pull request #5 from MarlinFirmware/RCBugFix

Rc bug fix

Commit:
Merge pull request #6 from MarlinFirmware/RCBugFix

Rc bug fix

Commit:
BugFix and optimization

.) long to int: Extruder stalls at 10kHz / 20kHz step limits with long.
.) Take the delta_adv_steps calculation out of the step_loops loop. Wasted calculation performance if done inside.
.) >> 2 replaced by 3: Is divide by 8. Reason: Timer 0 runs at 16/8=2MHz, Timer 1 at 16/64=0.25MHz. ==> 2/0.25=8.

Commit:
Speed improvement

Extruder stalls at 10kHz/20kHz limits with long.

esenapaj commented on Marlin/stepper.cpp:
```diff
@@ -593,7 +595,7 @@ void Stepper::isr() {
 
     #if ENABLED(ADVANCE) || ENABLED(LIN_ADVANCE)
-      eISR_Rate = (timer >> 2) * step_loops / abs(e_steps[TOOL_E_INDEX]);
+      eISR_Rate = (timer >> 3) * step_loops / abs(e_steps[TOOL_E_INDEX]); //>> 3 is divide by 8. Reason: Timer 0 runs at 16/8=2MHz, Timer 1 at 16/64=0.25MHz. ==> 2/0.25=8.
```
Is "Timer 1 runs at 16/8=2MHz, Timer 0 at 16/64=0.25MHz" right?

Because,
`Stepper::isr()` ... `TIMER1_COMPA_vect`
`Stepper::advance_isr()` ... `TIMER0_COMPA_vect`

Sebastianv650 replied:
Yes of course, typo :-/

Sebastianv650 commented:
Hm, I'm not 100% sure any more at which frequency each timer is running and if the extruder-ISR interferes with the temperature-ISR. Can someone please confirm if I got it right or not:
- Stepper ISR is running at TIMER1_COMPA
- Stepper extruder ISR is running at TIMER0_COMPA
- Temperature ISR is running at TIMER0_COMPB

In the LIN_ADVANCE timer code I assume that Timer 0 is 1/8 as fast as Timer 1.
Inside init() of the Stepper.cpp Timer 1 is set to CTC mode and a divider of 8.
Inside init() of Temperature.cpp only the Output compare register B is set to 128 and the interrupt is enabled. So the rest of the registers should be at Arduino defaults, as defined in wiring.c - which means the timer is set to Fast-PWM mode and a divider of 64.

BUT: The comment above the Temperature ISR says:

``` cpp
// This ISR uses the compare method so it runs at the base
// frequency (16 MHz / **256** = 62500 Hz), but at the TCNT0 set
// in OCR0B above (128 or halfway between OVFs).
```

So **question 1** is: At which speed is Timer 0 running? I would say 250kHz, the comment says 62.5kHz. If it's 62.5kHz, my LIN_ADVANCE code is faulty. If it's 250kHz, the Temperature ISR will be called 4x more often than intended by the programmer who has written the comment, meaning we are wasting CPU cycles with maybe not useful temp-checks.

**Question 2**: When LIN_ADVANCE or ADVANCE is enabled, inside init() of the Stepper.cpp Timer 0 is set from Fast-PWM mode to Normal mode. As far as I understand Fast_PWM, it has nothing todo with the interrupt frequency. So it shouldn't affect the Temperature ISR, true?

thinkyhead commented:
> At which speed is Timer 0 running?

The timer counts up at a rate of 250KHz using the default prescaler 1/64.

> If it's 250kHz, the Temperature ISR will be called 4x more often than intended by the programmer who has written the comment

The comment is mistaken. Since the timer runs at 250KHz but the isr is called on only every 256th iteration, then the actual Temperature ISR frequency is 976.5625Hz, not 62.5KHz.

thinkyhead commented:
> As far as I understand Fast_PWM, it has nothing todo with the interrupt frequency.

Correct. The mode of the timer is changed, so it will be used for `ISR(TIMER0_COMPA_vect)`, but the prescaler is not changed, so the timer continues to run at 250KHz.

thinkyhead commented:
> In the LIN_ADVANCE timer code I assume that Timer 0 is 1/8 as fast as Timer 1.

Correct. Timer 1 uses a prescaler of 8 (2MHz) versus Timer 0 which uses 64 (250KHz).

thinkyhead commented:
I've done a rebase and squash at #4980. I'm not entirely sure that moving the `delta_adv_steps` code out of the `step_loops` loop was a good idea. The `step_loops` loop is supposed to apply up to 4 steps worth of change to each variable involved in stepping. If the `delta_adv_steps` code is meant to apply to only a single step, then it should stay inside the loop.

Sebastianv650 commented:
`delta_adv_steps` is independent from the amount of `step_loops` done at once. It's only adding the `delta_adv_steps` based on the extrusion speed change since the last ISR loop.
If it is calculated 2 or 4 times in the loop, the first run will give the `delta_adv_steps`, every next run will result in `delta_adv_steps`=0.

thinkyhead commented:
> If it is calculated 2 or 4 times in the loop, the first run will give the delta_adv_steps, every next run will result in delta_adv_steps=0.

Makes sense, and good to know! Do you think I should prime the mixing steppers using the whole number of steps, or (as it is now) proportionally? I made the change below:

``` cpp
#if ENABLED(LIN_ADVANCE)
  if (current_block->use_advance_lead) {
    int delta_adv_steps = (((long)extruder_advance_k * current_estep_rate[TOOL_E_INDEX]) >> 9);
    #if ENABLED(MIXING_EXTRUDER)
      // Mixing extruders apply advance lead proportionally
      MIXING_STEPPERS_LOOP(j) delta_adv_steps -= current_adv_steps[j];
      MIXING_STEPPERS_LOOP(j) {
        int steps = delta_adv_steps * current_block->step_event_count / current_block->mix_event_count[j];
        e_steps[j] += steps;
        current_adv_steps[j] += steps;
      }
    #else
      // For most extruders, advance the single E stepper
      delta_adv_steps -= current_adv_steps[TOOL_E_INDEX];
      e_steps[TOOL_E_INDEX] += delta_adv_steps;
      current_adv_steps[TOOL_E_INDEX] += delta_adv_steps;
    #endif
 }
#endif
```

Particularly:

``` cpp
MIXING_STEPPERS_LOOP(j) delta_adv_steps -= current_adv_steps[j];
```

All the stepper's steps need to be included because the steps are spread out, but they add up to around equal to the value of `1.0 * current_adv_steps[TOOL_E_INDEX]` (i.e., the tool-based version).

If your `current_adv_steps` variable only needs only to be tool-based, and has no bearing later –on step counts for example– then I can modify the placement…

``` cpp
#if ENABLED(LIN_ADVANCE)
  if (current_block->use_advance_lead) {
    int delta_adv_steps = (((long)extruder_advance_k * current_estep_rate[TOOL_E_INDEX]) >> 9) - current_adv_steps[TOOL_E_INDEX];
    current_adv_steps[TOOL_E_INDEX] += delta_adv_steps;
    #if ENABLED(MIXING_EXTRUDER)
      // Mixing extruders apply advance lead proportionally
      MIXING_STEPPERS_LOOP(j)
        e_steps[j] += delta_adv_steps * current_block->step_event_count / current_block->mix_event_count[j];
    #else
      // For most extruders, advance the single E stepper
      e_steps[TOOL_E_INDEX] += delta_adv_steps;
    #endif
 }
#endif
```

Sebastianv650 commented:
Let's see if my answer is answering your question:

`current_adv_steps` is only needed per physical extruder. I don't know where the "j" comes from, but I guess it stands for a virtual extruder. Lets say we have 3 extruders "R", "G", "B" and 2 virtual ones we use for printing:
- 1 is 50% "R" and 50% "G"
- 2 is 50% "R" and 50% "B".
  We have to track the pre-tension (`current_adv_steps`) inside every filament from R, G and B. If we print a fast perimeter with virtual extruder 1 which leads to 20 `current_adv_steps` total, they will distribute 10 `current_adv_steps` for R and 10 `current_adv_steps` for G. B is unused and `current_adv_steps` = 0.
  If we now print with virtual extruder 2 at the same speed, R will not get any more `current_adv_steps` because it's pre-tension is already 10 steps (10 needed - 10 `current_adv_steps`). But the 10 for B will be added (10 steps - 0 `current_adv_steps`).

By the way, I just found this a few lines lower in the stepper.cpp:

``` cpp
    #if ENABLED(LIN_ADVANCE)

      if (current_block->use_advance_lead)
        current_estep_rate[TOOL_E_INDEX] = ((uint32_t)acc_step_rate * current_block->e_speed_multiplier8) >> 8;

      if (current_block->use_advance_lead) {
        #if ENABLED(MIXING_EXTRUDER)
          MIXING_STEPPERS_LOOP(j)
            current_estep_rate[j] = ((uint32_t)acc_step_rate * current_block->e_speed_multiplier8 * current_block->step_event_count / current_block->mix_event_count[j]) >> 8;
        #else
          current_estep_rate[TOOL_E_INDEX] = ((uint32_t)acc_step_rate * current_block->e_speed_multiplier8) >> 8;
        #endif
      }

    #elif ENABLED(ADVANCE)
```

It looks like there was a mistake during the `MIXING_EXTRUDER` implementation, the first `if (current_block->use_advance_lead)` should be deleted as it is done in the #else block.

Last one, you are calculating a `current_estep_rate[j]` for acceleration and deceleration, but not in cruising state?

thinkyhead commented:
> It looks like there was a mistake during the `MIXING_EXTRUDER` implementation

Oops… Patched!

thinkyhead commented:
>  I don't know where the "j" comes from, but I guess it stands for a virtual extruder.

`MIXING_STEPPERS_LOOP(j)` loops through the physical steppers, all of which belong to the one extruder. Perhaps in future I will write up some good documentation on how I'm adapting the stepper code for different kinds of extruders.

thinkyhead commented:
> you are calculating a `current_estep_rate[j]` for acceleration and deceleration, but not in cruising state?

Well, I've tried not to be the author of anything too original in the stepper code. So if that's how it works now, then I assume that's how it has been. But, is there some case where the `current_estep_rate` is never calculated, because there is no acceleration on a move? Because otherwise, it will always end up being set to the "cruising rate" at the last accelerated step. Once it's been calculated at the end of acceleration, there's no need to keep calculating it while the rate is remaining constant.
